### PR TITLE
[AN-4369][AN-4368][AN-4364][AN-4385] Separate camera hardware logic from controller logic and add tests for the controller

### PR DESCRIPTION
**Pull request by deanrobertcook**
This PR moves any direct access to the camera API to a new class which can now be mocked, allowing the logic in the controller itself to be tested. I also included a potential fix for one of the Hockey app crashes, where I was restarting the camera without trying and catching any potential errors.

Another improvement includes returning the supported flash modes for the current camera immediately in the `onCameraLoaded()` callback, so that outside classes don't have to call the controller and risk race conditions. 

For now I included just some of the most straight-forward tests. As I figure out some other problems, I'll hopefully include some more if needed. 

#### Test Overview:
The only thing that can really be manually tested is ticket AN-4385. Everything else will have to be monitored over hockey in the coming weeks. 

#### Tickets:
https://wearezeta.atlassian.net/browse/AN-4369
https://wearezeta.atlassian.net/browse/AN-4368
https://wearezeta.atlassian.net/browse/AN-4364
https://wearezeta.atlassian.net/browse/AN-4385

#### APK
[Download build #7367](http://192.168.10.18:8080/job/Pull%20Request%20Builder/7367/artifact/build/artifact/wire-dev-PR49-7367.apk)
[Download build #7368](http://192.168.10.18:8080/job/Pull%20Request%20Builder/7368/artifact/build/artifact/wire-dev-PR49-7368.apk)
[Download build #7369](http://192.168.10.18:8080/job/Pull%20Request%20Builder/7369/artifact/build/artifact/wire-dev-PR49-7369.apk)
[Download build #7371](http://192.168.10.18:8080/job/Pull%20Request%20Builder/7371/artifact/build/artifact/wire-dev-PR49-7371.apk)
[Download build #7373](http://192.168.10.18:8080/job/Pull%20Request%20Builder/7373/artifact/build/artifact/wire-dev-PR49-7373.apk)
[Download build #7374](http://192.168.10.18:8080/job/Pull%20Request%20Builder/7374/artifact/build/artifact/wire-dev-PR49-7374.apk)
[Download build #7377](http://192.168.10.18:8080/job/Pull%20Request%20Builder/7377/artifact/build/artifact/wire-dev-PR49-7377.apk)
[Download build #7381](http://192.168.10.18:8080/job/Pull%20Request%20Builder/7381/artifact/build/artifact/wire-dev-PR49-7381.apk)
[Download build #7391](http://192.168.10.18:8080/job/Pull%20Request%20Builder/7391/artifact/build/artifact/wire-dev-PR49-7391.apk)
[Download build #7392](http://192.168.10.18:8080/job/Pull%20Request%20Builder/7392/artifact/build/artifact/wire-dev-PR49-7392.apk)
[Download build #7393](http://192.168.10.18:8080/job/Pull%20Request%20Builder/7393/artifact/build/artifact/wire-dev-PR49-7393.apk)

### CODE REVIEW

File: app/src/main/scala/com/waz/zclient/camera/GlobalCameraController.scala [hunk 1]
```diff
@@ -34,24 +38,121 @@ import timber.log.Timber
...
+  def setFocusArea(touchRect: Rect, w: Int, h: Int) = {
+    val promise = Promise[Unit]()
+    Future {
+      currentCamera match {
+        case Some(c) => c.setFocusArea(touchRect, w, h).future.onComplete {
+          case Success(_) => promise.success(())
+          case Failure(e) =>
+            Timber.w(e, "Autofocus failed")
+            promise.failure(e)
+        }
```
Comment (zbsz) [inline app/src/main/scala/com/waz/zclient/camera/GlobalCameraController.scala]: `completeWith`

File: app/src/main/scala/com/waz/zclient/camera/GlobalCameraController.scala [hunk 2]
```diff
@@ -62,16 +163,31 @@ class GlobalCameraController(cxt: WireContext)(implicit eventContext: EventConte
...
       Seq.empty
   }
+}
 
-  @volatile private var currentCamera = Option.empty[Camera]
+trait WireCamera {
+  def getPreviewSize: PreviewSize
 
-  private var loadFuture = CancellableFuture.cancelled[PreviewSize]()
+  def takePicture(shutter: => Unit): Promise[Array[Byte]]
```
Comment (zbsz) [inline app/src/main/scala/com/waz/zclient/camera/GlobalCameraController.scala]: Actually not sure if returning `Promise` is good here. `Promise` is mutable, caller can modify it (calling `success` of `failure`), you probably don't want that to happen. Also having more general `Future` in api would give you more felxibility.

File: app/src/main/scala/com/waz/zclient/camera/GlobalCameraController.scala [hunk 3]
```diff
@@ -98,159 +214,120 @@ class GlobalCameraController(cxt: WireContext)(implicit eventContext: EventConte
...
+          ), focusWeight)
+
+          setParams(_.setFocusAreas(List(focusArea).asJava))
+          if (!settingFocus) try {
+            settingFocus = true
             c.autoFocus(new AutoFocusCallback {
               override def onAutoFocus(s: Boolean, cam: Camera) = {
                 if (!s) Timber.w("Focus was unsuccessful - ignoring")
                 promise.success(())
+                settingFocus = false
```
Comment (zbsz) [inline app/src/main/scala/com/waz/zclient/camera/GlobalCameraController.scala]: Is it specified on which thread `onAutoFocus` callback is executed? We probably should use `volatile` annotation for `settingFocus`.

**Comment by mykola-mokhnach** [comment]
I'm unable to find any visible issues with camera view

File: app/src/main/scala/com/waz/zclient/camera/GlobalCameraController.scala (hunk 1 above)
Comment (zbsz) [inline app/src/main/scala/com/waz/zclient/camera/GlobalCameraController.scala]: Or get rid of `promise` and use `Future.flatMap` or 'flatten'

File: app/src/main/scala/com/waz/zclient/camera/GlobalCameraController.scala (hunk 3 above)
Comment (deanrobertcook) [inline app/src/main/scala/com/waz/zclient/camera/GlobalCameraController.scala]: done

**Pull request merged**